jme3-core/src/main/java/com/jme3/anim/AnimLayer.java

```diff
@@ -0,0 +1,225 @@
+     * layer, or null if unknown.
+     */
+    private Object manager;
+
+    /**
+     * Instantiates a layer without a manager or a current Action, owned by the
+     * specified composer.
+     *
+     * @param composer the owner (not null, alias created)
+     * @param mask the AnimationMask (alias created) or null to allow this layer
+     *     to animate the entire model
+     */
+    AnimLayer(AnimComposer composer, AnimationMask mask) {
+        assert composer != null;
+        this.composer = composer;
+
+        this.mask = mask;
+    }
+
+    /**
+     * Returns the Action that's currently running.
+     *
+     * @return the pre-existing instance, or null if none
+     */
+    public Action getCurrentAction() {
+        return currentAction;
+    }
+
+    /**
+     * Returns the current manager.
+     *
+     * @return the pre-existing object (such as an AnimEvent) or null for
+     *     unknown
+     */
+    public Object getManager() {
+        return manager;
+    }
+
+    /**
+     * Returns the animation mask.
+     *
+     * @return the pre-existing instance, or null if this layer can animate the
+     *     entire model
+     */
+    public AnimationMask getMask() {
+        return mask;
+    }
+
+    /**
+     * Returns the animation time, in scaled seconds.
+     *
+     * @return the current animation time (not negative)
+     */
+    public double getTime() {
+        return time;
+    }
+
+    /**
+     * Runs the specified Action, starting from time = 0. This cancels any
+     * Action previously running on this layer.
+     *
+     * @param actionToRun the Action to run (alias created) or null for no
+     *     action
+     */
+    public void setCurrentAction(Action actionToRun) {
+        this.time = 0.0;
+        this.currentAction = actionToRun;
+    }
+
+    /**
+     * Assigns the specified manager. This cancels any manager previously
+     * assigned.
+     *
+     * @param desiredManager the desired manager (such as an AnimEvent, alias
+     *     created) or null for unknown manager
+     */
+    public void setManager(Object desiredManager) {
+        this.manager = desiredManager;
+    }
+
+    /**
+     * Changes the animation time, wrapping the specified time to fit the
+     * current Action. An Action be running.
+     *
+     * @param desiredAnimationTime the desired time (in scaled seconds)
+     */
+    public void setTime(double desiredAnimationTime) {
+        double length = currentAction.getLength();
+        if (desiredAnimationTime >= 0.0) {
+            time = desiredAnimationTime % length;
+        } else {
+            time = (desiredAnimationTime % length) + length;
+        }
+    }
+
+    /**
+     * Updates the animation time and the current Action during a
+     * controlUpdate().
+     *
+     * @param realSeconds the amount to advance the current Action, in real
+     *     seconds
+     */
+    void update(float realSeconds) {
```

Comment:
it would be more intuitive to keep the variable name tpf that is passed by the controlUpdate() method.

change parameter name realSeconds -> tpf

Reply:
"tpf" assumes the reader knows what those initials stand for, doesn't communicate the units, and assumes that the method will be called once per frame. Doesn't seem more intuitive to me.

Reply:
I always thought that the controlUpdate() method was invoked once per frame, didn't it? The javadoc already describes all this information very well. We could add: `@param tpf (time per frame) the amount to advance the current Action, in real seconds`

Reply:
It doesn't seem strange to me. The names of formal arguments should make sense in the context where they're seen. In the context of a scene-graph control, "frame" usually means a render cycle. In the context of an animation action, "frame" usually means keyframe, so any mention of "frame" here could cause confusion.

I try not to rely much on javadoc to convey what methods and args do. Many dev tools provide formal argument names without javadoc, so devs often make assumptions about what methods and arguments do based on their names, leading to mistakes like [this one](https://hub.jmonkeyengine.org/t/solved-charactercontrol-update-not-called-automatically/45101).